## Tarea Práctica 1
### Álgebra Lineal y Optimización para Data Science 2023/2

#### Integrantes: Sebastián Dinator, Joaquín Olivares, Benjamín Valdés.

### Ejercicio 3)

In [1]:
from ucimlrepo import fetch_ucirepo

# Fetch Glass Identification dataset
glass_identification = fetch_ucirepo(id=42)

# Extract features and targets
X = glass_identification.data.features
y = glass_identification.data.targets

# Explorar el conjunto de datos (opcional)
import pandas as pd
df = pd.DataFrame(data=X, columns=glass_identification.data.feature_names)
df['Target'] = y
df.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Target
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0       1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0       1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0       1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0       1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0       1

#### Ejercicio 3B)

In [13]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
import matplotlib.pyplot as plt

# Dividir el conjunto de datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Entrenar el modelo LDA (sin probabilidades a priori)
lda = LinearDiscriminantAnalysis()
lda.fit(X_train, y_train)

# Predecir etiquetas de clase para el conjunto de prueba
y_pred_lda = lda.predict(X_test)

# Calcular precisión y matriz de confusión para LDA
accuracy_lda = accuracy_score(y_test, y_pred_lda)
confusion_matrix_lda = confusion_matrix(y_test, y_pred_lda)

# Imprimir métricas para LDA
print("Precisión del modelo LDA:", accuracy_lda)
print("Matriz de Confusión para LDA:\n", confusion_matrix_lda)


Precisión del modelo LDA: 0.6461538461538462
Matriz de Confusión para LDA:
 [[16  2  1  0  0  0]
 [ 6 15  1  0  1  0]
 [ 3  1  0  0  0  0]
 [ 0  2  1  1  2  0]
 [ 0  0  0  0  2  1]
 [ 1  0  0  0  1  8]]


C:\Users\sebad\anaconda3\Lib\site-packages\sklearn\utils\validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


La precisión del modelo LDA es aproximadamente de un 64.6% de las clasificaciones de nuestro conjunto de prueba.

#### Ejercicio 3C)

In [9]:
from sklearn.mixture import GaussianMixture

# Entrenar el modelo de Gaussian Mixture
gmm = GaussianMixture(n_components=2)  # Cambiar el número de componentes según sea necesario
gmm.fit(X_train)

# Predecir etiquetas de clase para el conjunto de prueba
y_pred_gmm = gmm.predict(X_test)

# Calcular precisión y matriz de confusión para GMM
accuracy_gmm = accuracy_score(y_test, y_pred_gmm)
confusion_matrix_gmm = confusion_matrix(y_test, y_pred_gmm)

# Imprimir métricas para GMM
print("Precisión del modelo GMM:", accuracy_gmm)
print("Matriz de Confusión para GMM:\n", confusion_matrix_gmm)

Precisión del modelo GMM: 0.2923076923076923
Matriz de Confusión para GMM:
 [[ 0  0  0  0  0  0  0]
 [ 0 19  0  0  0  0  0]
 [ 7 16  0  0  0  0  0]
 [ 1  3  0  0  0  0  0]
 [ 6  0  0  0  0  0  0]
 [ 3  0  0  0  0  0  0]
 [10  0  0  0  0  0  0]]


C:\Users\sebad\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1436: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


La precisión del modelo GMM es mucho más baja que la de LDA, con aproximadamente un 29.2%. Se predijo sólo una clase de todas las muestras del conjunto de prueba, por lo que se puede asumir que no pudo capturar la estructura de los datos correctamente.

Los resultados de LDA son mucho mejores que los de GMM, ya que LDA es un modelo de clasificación lineal, mientras que GMM es un modelo de clasificación no lineal. Por lo tanto, LDA es capaz de capturar la estructura de los datos, mientras que GMM no.

#### Ejercicio 3D)

In [10]:
import numpy as np

# Generar 10 nuevas observaciones de prueba (cambiar según tus necesidades)
new_samples = np.random.rand(10, X.shape[1])

# Clasificar con LDA
y_new_pred_lda = lda.predict(new_samples)

# Clasificar con GMM
y_new_pred_gmm = gmm.predict(new_samples)

# Imprimir las predicciones
print("Predicciones con LDA para 10 nuevas observaciones:", y_new_pred_lda)
print("Predicciones con GMM para 10 nuevas observaciones:", y_new_pred_gmm)

Predicciones con LDA para 10 nuevas observaciones: [3 3 3 3 3 3 3 3 3 3]
Predicciones con GMM para 10 nuevas observaciones: [1 1 1 1 1 1 1 1 1 1]


C:\Users\sebad\anaconda3\Lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but LinearDiscriminantAnalysis was fitted with feature names
  warnings.warn(
C:\Users\sebad\anaconda3\Lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but GaussianMixture was fitted with feature names
  warnings.warn(


Tanto el modelo LDA como GMM asignan la misma etiqueta a todas las observaciones nuevas, por lo que se puede apreciar que ambos modelos no son capaces de clasificar correctamente las observaciones nuevas. Esto puede deberse a que los modelos no son capaces de capturar la estructura de los datos, o que los datos no son linealmente separables.